Lib/fractions.py

```diff
@@ -380,32 +384,110 @@ def reverse(b, a):
-        return Fraction(a.numerator * db + b.numerator * da,
-                        da * db)
+    # Rational arithmetic algorithms: Knuth, TAOCP, Volume 2, 4.5.1.
+    #
+    # Assume input fractions a and b are normalized.
+    #
+    # 1) Consider addition/substraction.
+    #
+    # Let g = gcd(da, db). Then
+    #
+    #              na   nb    na*db ± nb*da
+    #     a ± b == -- ± -- == ------------- ==
+    #              da   db        da*db
+    #
+    #              na*(db//g) ± nb*(da//g)    t
+    #           == ----------------------- == -
+    #                      (da*db)//g         d
+    #
+    # Now, if g > 1, we're working with smaller integers.
+    #
+    # Note, that t, (da//g) and (db//g) are pairwise coprime.
+    #
+    # Indeed, (da//g) and (db//g) share no common factors (they were
+    # removed) and da is coprime with na (since input fractions are
+    # normalized), hence (da//g) and na are coprime.  By symmetry,
+    # (db//g) and nb are coprime too.  Then,
+    #
+    #     gcd(t, da//g) == gcd(na*(db//g), da//g) == 1
+    #     gcd(t, db//g) == gcd(nb*(da//g), db//g) == 1
+    #
+    # Above allows us optimize reduction of the result to lowest
+    # terms.  Indeed,
+    #
+    #     g2 = gcd(t, d) == gcd(t, (da//g)*(db//g)*g) == gcd(t, g)
+    #
+    #                       t//g2                   t//g2
+    #     a ± b == ----------------------- == ----------------
+    #              (da//g)*(db//g)*(g//g2)    (da//g)*(db//g2)
+    #
+    # is a normalized fraction.  This is useful because the unnormalized
+    # denominator d could be much larger than g.
+    #
+    # We should special-case g == 1, since 60.8% of randomly-chosen
+    # integers are coprime:
+    # https://en.wikipedia.org/wiki/Coprime_integers#Probability_of_coprimality
+    #
+    # 2) Consider multiplication
+    #
+    # Let g1 = gcd(na, db) and g2 = gcd(nb, da), then
+    #
+    #            na*nb    na*nb    (na//g1)*(nb//g2)
+    #     a*b == ----- == ----- == -----------------
+    #            da*db    db*da    (db//g1)*(da//g2)
+    #
+    # Note, that after divisions we're multiplying smaller integers.
+    #
+    # Also, the resulting fraction is normalized, because each of
+    # two factors in the numerator is coprime to each of the two factors
+    # in the denominator.
+    #
+    # Indeed, pick (na//g1).  It's coprime with (da//g2), because input
+    # fractions are normalized.  It's also coprime with (db//g1), because
+    # common factors are removed by g1 == gcd(na, db).
+
+    def _add_sub_(a, b, pm=int.__add__):
+        na, da = a._numerator, a._denominator
+        nb, db = b._numerator, b._denominator
+        g = math.gcd(da, db)
+        if g == 1:
+            return Fraction(pm(na * db, da * nb), da * db, _normalize=False)
+        else:
+            s = da // g
+            t = pm(na * (db // g), nb * s)
+            g2 = math.gcd(t, g)
+            return Fraction(t // g2, s * (db // g2), _normalize=False)
```

Comment:
```suggestion
            if g2 == 1:
                return Fraction(t , s * db), _normalize=False)
            else:
                return Fraction(t // g2, s * (db // g2), _normalize=False)
```
Suggested for the reason `g == 1` is a special case in the enclosing block: 1 is most likely, and some work can be skipped when the gcd is 1. I'm not aware of a proof that `g2` is most often 1, but trying some millions of "random" inputs in a test driver it was. As a sanity check, I looked at gmp's implementation, and saw that it special-cases 1 here too.

Reply:
Yes, I saw that, but I don't see any validation for this case in references.

I suspect, the reason may be same as for `g == 1` case, i.e. probability of coprimes for random inputs (`t` and `gcd` seems to be independent).  I dig into the history and that indeed was [the argument](https://gmplib.org/repo/gmp/rev/177ba6188ca8) to introduce this special case.  But that's not a real proof.  Either we should include such one or find a better reference to defend this change.  Or it's ok if we just point to [aors.c](https://gmplib.org/repo/gmp/file/tip/mpq/aors.c) (that I already did in the commit)?

Reply:
> `t` and `gcd` seems to be independent

But in fact...  We can change both parts independently (e.g. alter input numerators and freeze denominators).  Thus, it seems we can use same argument as for `g == 1`.

Reply:
There is no possible question about correctness here:  if `g2 == 1`, it's utterly obvious that there's no need to explicitly divide by `g2`.

So it's a probabilistic optimization, and doesn't need "proof". It depends on the data. As I've noted before, the idea that the top-level `g == 1` test is usually true is plain false in _some_ contexts (e.g., for any addition derived from non-integer floats with different denominators, the gcd is always the smaller denominator - they're all exact powers of 2).

It's plenty good enough for me that, as I said, `g2` was in fact 1 in the majority of some millions of "random" input cases I tried. That's all the justification a heuristic needs, and especially in a context where the test (`g2 == 1`) is a trivial cost compared to the potential savings. It doesn't matter if `g2` has a trillion bits - that test is cheap.

Reply:
> There is no possible question about correctness

Sure, that wasn't my point.

> So it's a probabilistic optimization, and doesn't need "proof".

I don't think so.  At least we should provide an example of some contexts, where it does make sense.  Or (better) - show, that it will payoff for random inputs.

Reply:
For the third :wink: time, I already tried millions of "random" inputs  I have no interest in spending more time on it, because the eventual conclusion is already obvious.

If you need a specific context, I already gave one, but didn't spell out the consequences: additions derived from non-integer floats. There `g == 1` is never true, but `g2 == 1` is always true.

    odd1 / 2**i + odd2 / 2**j  and assume 0 < i < j
    then g = 2**i
    the numerator is  odd1 * 2**(j-i) + odd2, which is necessarily odd
    so g2 = gcd(numerator, g) = gcd(odd, power_of_2) is necessarily 1

Reply:
Ok, @tim-one, are you ok with the documentation addition, that argue this change?

Reply:
BTW, [same](https://github.com/python/cpython/pull/24779#issuecomment-792227837) random, unrelated test failure.

Reply:
> are you ok with the documentation addition, that argue this change?

Oh yes. They're a real help - thanks!

Reply:
Maybe then it worth special-case `g1` and `g2` for `_mul()` as well (like [this](https://bugs.python.org/issue43420#msg388212))?

Such checks, unfortunately, are costly for very small components (that was the reason I did only `g==1` optimization in `_add_sub_()`).  Same could be seen for Raymond's benchmarks in the issue thread.

Reply:
> Maybe then it worth special-case `g1` and `g2` for `_mul()` as well (like [this](https://bugs.python.org/issue43420#msg388212))?

I think so, yes. There's every reason to expect these gcds will usually be 1 (indeed, for multiplying two rationals obtained from floats, they're always 1).

The overhead of running Python-level code swamps the actual arithmetic needed for rationals with small components, but so it goes. That's not "a reason", to me, to avoid good algorithms with overheads - it's a reason not to use fractions.py at all in code for which the speed of rationals with small components is important.

Reply:
Ok, done.  Comments were slightly expanded also for the case of fractions, obtained from floats.